addons/binding/org.openhab.binding.ipp/src/main/java/org/openhab/binding/ipp/internal/discovery/IppPrinterDiscoveryParticipant.java

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (c) 2015 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.ipp.internal.discovery;
+
+import java.net.InetAddress;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import javax.jmdns.ServiceInfo;
+
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.io.transport.mdns.discovery.MDNSDiscoveryParticipant;
+import org.openhab.binding.ipp.IppBindingConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * discovers ipp printers announced by mDNS
+ * 
+ * @author Tobias Bräutigam
+ *
+ */
+public class IppPrinterDiscoveryParticipant implements MDNSDiscoveryParticipant {
+
+	private Logger logger = LoggerFactory
+			.getLogger(IppPrinterDiscoveryParticipant.class);
+
+	@Override
+	public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+		return Collections.singleton(IppBindingConstants.PRINTER_THING_TYPE);
+	}
+
+	@Override
+	public String getServiceType() {
+		return "_ipp._tcp.local.";
+	}
+
+	@Override
+	public ThingUID getThingUID(ServiceInfo service) {
+		if (service != null) {
+			logger.trace("ServiceInfo: {}", service);
+			if (service.getType() != null) {
+				if (service.getType().equals(getServiceType())) {
+					String uidName = getUIDName(service);
+					return new ThingUID(
+							IppBindingConstants.PRINTER_THING_TYPE,
+							uidName);
+				}
+			}
+		}
+		return null;
+	}
+	
+	private String getUIDName(ServiceInfo service) {
+		return service.getName().replaceAll("[^A-Za-z0-9_]", "_");
+	}
+	
+	private InetAddress getIpAddress(ServiceInfo service) {
+		InetAddress address = null;
+		for (InetAddress addr : service.getInet4Addresses()) {
+			return addr;
+		}
+		// Fallback for Inet6addresses
+		for (InetAddress addr : service.getInet6Addresses()) {
+			return addr;
+		}
+		return address;
+	}
+	
+	private String getPrinterName(ServiceInfo service) {
+		String rp = service.getPropertyString("rp");
+		if (rp==null) {
```

Comment:
Is "rp" really a good default? In my case, "rp" is filled ("printers/HP_LaserJet_M2727nf_MFP__745B02_"), but the service name ("HP LaserJet M2727nf MFP (745B02) @ Kona") would actually be a much better label text. Did you come across situations, where there is a name given, but which is not usable as a label?